core/src/rpc.rs

```diff
@@ -382,6 +383,94 @@ impl JsonRpcRequestProcessor {
+        let slot = self.get_slot(Some(CommitmentConfig::finalized()));
+        let first_available_block = self.get_first_available_block();
+
+        let mut epoch = epoch_schedule.get_epoch_and_slot_index(slot).0;
+        let mut epoch_info: Option<(Slot, UnixTimestamp, solana_transaction_status::Rewards)> =
+            None;
+
+        while epoch > lowest_epoch {
+            let first_slot_in_epoch = epoch_schedule.get_first_slot_in_epoch(epoch);
+            if first_slot_in_epoch < first_available_block {
+                // RPC node is out of history data
+                break;
+            }
+
+            let first_confirmed_block_in_epoch = *self
+                .get_confirmed_blocks_with_limit(first_slot_in_epoch, 1, None)?
+                .get(0)
+                .ok_or(RpcCustomError::BlockNotAvailable {
+                    slot: first_slot_in_epoch,
+                })?;
+
+            let first_confirmed_block = if let Ok(first_confirmed_block) =
+                self.get_confirmed_block_at_slot(first_confirmed_block_in_epoch, None)
+            {
+                first_confirmed_block
+            } else {
+                break; // doesn't have this block
+            };
+
+            let epoch_start_time = if let Some(block_time) = first_confirmed_block.block_time {
+                block_time
+            } else {
+                break;
+            };
+
+            // Rewards for the previous epoch are found in the first confirmed block of the current epoch
+            let previous_epoch_rewards = first_confirmed_block.rewards;
+
+            if let Some((effective_slot, epoch_end_time, epoch_rewards)) = epoch_info {
+                let epoch_duration = if epoch_end_time > epoch_start_time {
+                    Some((epoch_end_time - epoch_start_time) as f64)
+                } else {
+                    None
+                };
+
+                if let Some(reward) = epoch_rewards
+                    .into_iter()
+                    .find(|reward| reward.pubkey == pubkey.to_string())
+                {
+                    if reward.post_balance > reward.lamports.try_into().unwrap() {
+                        let rate_change = reward.lamports.abs() as f64
+                            / (reward.post_balance as f64 - reward.lamports as f64);
+
+                        let apr = epoch_duration.map(|epoch_duration| {
+                            let epochs_per_year = (SECONDS_PER_DAY * 356) as f64 / epoch_duration;
+                            rate_change * epochs_per_year
+                        });
+
+                        all_epoch_rewards.push(RpcEpochReward {
+                            epoch,
+                            effective_slot,
+                            amount: reward.lamports.abs() as u64,
+                            post_balance: reward.post_balance,
+                            percent_change: rate_change * 100.0,
+                            apr: apr.map(|r| r * 100.0),
```

Comment:
The client can compute `percent_change` and `apr` if they want, let's keep it out of the API

Reply:
Sounds good!